src/progress_bar.rs

```diff
@@ -638,22 +640,18 @@ impl MultiProgress {
 
     /// Modify the state on inserting/adding a progress bar
     fn extend_state(state: &mut MultiProgressState) -> usize {
-        let object_idx;
+        let new_obj = Some(MultiObject {
+            done: false,
+            draw_state: None,
+        });
         if let Some(i) = state.free_set.iter().next().cloned() {
```

Comment:
Hm, otherwise types aren't line up and we have to do like this:
```
        *(if let Some(i) = state.free_set.iter().next() {
            state.objects[*i] = new_obj;
            state.free_set.remove(i);
            i
        } else {
            state.objects.push(new_obj);
            state.objects.len() - 1
        })
```

Reply:
So worst case we can use `copied()` here instead of `cloned()`. But the whole `let i = ..iter().next(); ...remove(i);` pattern seems pretty ugly if we can also pick a data type that has a natural `pop()` method. Since `BTreeSet` doesn't have that, I still feel like `Vec` or `VecDeque` or `BinaryHeap` are more attractive.

Reply:
Nit: `cloned` for `Copy` types works exactly like `copied`: https://godbolt.org/z/1e3YaMnhx
Although clippy kept silent here despite on https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_copy

Reply:
Let's recap the requirements, the collection should have the following operation:
- add the new element to the collection (`insert(e)`)
- remove some element from the collection (`remove(e)`)
- get and remove _any_ existing number from the collection (`if let Some(e) = pop()`)

For me `HashSet` fulfils all the requirements with constant time for the some redundant memory consumption.
`BTreeSet` brings logarithmic time for all the operations and more overhead compared to `HashSet`, but given the bound fulfills the requirements as well.
`VecDeque` brings sequential time to the `remove`  operation, and doesn't have `remove(e)`, we have to use `retain(|x| x != e)` as the workaround.

Reply:
I just changed it to `Vec` after reviewing the requirements some more.